Src/PrimitiveUtilities.cpp

```diff
@@ -273,4 +273,167 @@ void Solver::createBox(const glm::vec3& translation, float scale, float k) {
+
+  // Add nodes in a grid
+  this->_nodes.reserve(
+      currentNodeCount + grid.width * grid.height);
+  for (uint32_t i = 0; i < grid.width; ++i) {
+    for (uint32_t j = 0; j < grid.height; ++j) {
+        uint32_t nodeId = grid.gridIdToNodeId(currentNodeCount, {i, j, 0});
+
+        Node& node = this->_nodes.emplace_back();
+        node.id = nodeId;
+        node.position = scale * glm::vec3(i, j, j) + translation;
+        node.prevPosition = node.position;
+        node.velocity = glm::vec3(0.0f);
+        node.mass = 1.0f;
+
+        // if (i == 0 && j == 0) {
+        //   this->_positionConstraints.push_back(
+        //       createPositionConstraint(this->_constraintId++, &node));
+        // }
+    }
+  }
+
+  // Add distance constraints in each grid cell
+  this->_distanceConstraints.reserve(
+      currentDistConstraintsCount +
+      8 * (grid.width - 1) * (grid.height - 1));
+  for (uint32_t i = 0; i < grid.width; ++i) {
+    for (uint32_t j = 0; j < grid.height; ++j) {
+        uint32_t node00 = grid.gridIdToNodeId(currentNodeCount, {i, j, 0});
+        uint32_t node01 = grid.gridIdToNodeId(currentNodeCount, {i, j + 1, 0});
+        uint32_t node10 = grid.gridIdToNodeId(currentNodeCount, {i + 1, j, 0});
+        uint32_t node11 = grid.gridIdToNodeId(currentNodeCount, {i + 1, j + 1, 0});
+
+        // Grid-aligned constraints
+        if (i < (grid.width - 1)) {
+            this->_distanceConstraints.push_back(createDistanceConstraint(
+                this->_constraintId++,
+                this->_nodes[node00],
+                this->_nodes[node10]));
+        }
+
+        if (j < (grid.height - 1)) {
+            this->_distanceConstraints.push_back(createDistanceConstraint(
+                this->_constraintId++,
+                this->_nodes[node00],
+                this->_nodes[node01]));
+        }
+
+        // Long diagonal constraints
+        if (i < (grid.width - 1) && j < (grid.height - 1)) {
+            this->_distanceConstraints.push_back(createDistanceConstraint(
+                this->_constraintId++,
+                this->_nodes[node00],
+                this->_nodes[node11]));
+        }
+    }
+  }
+
+  this->_triangles.reserve(
+      currentTriCount + (2 * grid.width * grid.height));
+  for (uint32_t i = 0; i < grid.width - 1; ++i) {
+    for (uint32_t j = 0; j < grid.height - 1; ++j) {
+      this->_triangles.push_back(
+          grid.gridIdToNodeId(currentNodeCount, {i, j, 0}));
+      this->_triangles.push_back(
+          grid.gridIdToNodeId(currentNodeCount, {i + 1, j, 0}));
+      this->_triangles.push_back(
+          grid.gridIdToNodeId(currentNodeCount, {i + 1, j + 1, 0}));
+
+      this->_triangles.push_back(
+          grid.gridIdToNodeId(currentNodeCount, {i, j, 0}));
+      this->_triangles.push_back(
+          grid.gridIdToNodeId(currentNodeCount, {i + 1, j + 1, 0}));
+      this->_triangles.push_back(
+          grid.gridIdToNodeId(currentNodeCount, {i, j + 1, 0}));
```

Comment:
This is no longer correct if we allow the vertices to contain a different count then the nodes list (e.g., adding floor vertices). While the Node IDs _should_ be used to specify constraints, the _vertex ID_ is what should be used to specify the indices for the traingles and/or lines. Again, this is because vertex IDs and node IDs won't correspond 1-to-1 if we have non-node vertices in the vertex buffer.

A hacky solution would be to use `currentVertexCount` instead of `currentNodeCount` here . But I think it would be better to have two vertex buffers, one for static objects (like the floor and imported static colliders), and one for all the nodes (dynamic objects).